design/animation.md

```diff
@@ -0,0 +1,355 @@
+		at runtime. See [References](#references).
+	3. An optional bezier easing function.
+
+```ts
+type Animatible = Actor | Material | AnimationInstance;
+type AnimatibleClassName = 'actor' | 'material' | 'animationinstance';
+type AnimatibleLike = ActorLike | MaterialLike | AnimationInstanceLike;
+type AnimationProp = Vector2 | Vector3 | Quaternion | Color3 | Color4 | number | string | boolean;
+type EasingFunction = [number, number, number, number];
+
+class AssetContainer {
+	public createAnimation(name: string, options: {
+		targets: { [name: string]: AnimatibleClassName },
+		tracks: Track<AnimationProp>[]
+	}): Animation;
+}
+class Animation {
+	id: Readonly<string>;
+	duration: Readonly<number>;
+	targets: { [name: string]: AnimatibleClassName };
+}
+
+type Track<T extends AnimationProp> = {
+	target: TargetPath<T>;
+	keyframes: Keyframe<T>[];
+}
+
+type Keyframe<T extends AnimationProp> = {
+	time: number;
+	value: T | TargetPath<T>;
+	relative?: boolean;
+	easing?: EasingFunction;
+}
+```
+
+### Paths
+
+Paths are references to properties on an object that can be animated (Actor, Material, etc.). Paths do not reference
+properties on a *particular* object, but instead reference a generic object of a given type by index. When bound,
+the animation will be given the list of objects which it will be modifying, and these path indices refer to that
+argument. Over the wire these are represented as strings, but are easily generated in the API with helper functions.
+
+```ts
+class TargetPath<T> {
+	private parent: TargetPath<any>;
+	constructor(name: string, children: Map<string, TargetPath<any>>) {
+		for (const [k,v] of children) {
+			this[k] = v;
+			v.parent = this;
+		}
+	}
+	public toJSON() {
+		// convert tree into string
+	}
+}
+function AnimateActor(name: string) {
+	return new TargetPath<never>(`actor.${name}`, {
+		transform: new TargetPath<never>("transform", {
+			local: new TargetPath<never>("local", {
+				position: new TargetPath<Vector3>("position", {
+					x: new TargetPath<number>("x"),
+					y: new TargetPath<number>("y"),
+					z: new TargetPath<number>("z")
+				})
+				// also rotation, scale
+			}),
+			app: new TargetPath<never>("app", {
+				/** lookAt replacement */
+				orientToward: new TargetPath<never>("orientToward", {
+					xy: new TargetPath<Vector3>("xy"),
+					y: new TargetPath<Vector3>("y")
+				})
+				// also position, rotation
+			})
+		})
+		// plus most other properties of actor
+	});
+}
+
+const track = {
+	target: AnimateActor("target").transform.local.position
+	// ...
+} as Track<Vector3>;
+```
+
+### References
+
+In addition to transitioning to a constant value on a property, you can also transition to a value on another object.
+This value is re-evaluated every frame on the client, so matches exactly.
+
+```ts
+assets.createAnimation({
+	targets: {
+		"start": "actor"
+		"end": "actor"
+	},
+	tracks: [{
+		target: AnimateActor("start").transform.local.position,
```

Comment:
Before implementing this part, let's discuss the format of the selector string. I want to make sure it is friendly to the addition of expressions later. I'm thinking something like:
* A reference with a binding name of "start": `"actor('bind:start', 'transform.local.position')"`
* A reference to a specific actor: `"actor('id:<guid>', 'transform.local.position')"`

This would be friendly to an expression parser.

Reply:
This is nice because it would be parsed as a function with parameters.